### PR TITLE
[script/pull_github_pr.sh] Check Gating status before merging

### PR DESCRIPTION
**Pull request by yaronkaikov**
Maintainers use scripts/pull_github_pr.sh from scylladb.git when merging PRs and before pushing to the next. We want to prevent merges from piling up on top of unstable builds. This change will check Gating's current status and notify the maintainers

Related to https://github.com/scylladb/scylla-pkg/issues/3644

Fixes: https://github.com/scylladb/scylladb/issues/20855

**merge script improvement, no need for backport**

### CODE REVIEW

File: scripts/pull_github_pr.sh
Comment (denesb) [inline scripts/pull_github_pr.sh:140]: Why not do this before the merge even happens? As-is, if the maintainer finds out the job is unstable, they have to undo the merge.

Comment (yaronkaikov) [reply scripts/pull_github_pr.sh:140]: Make sense. changed

**Comment by yaronkaikov** [comment]
v2:
- move `check_jenkins_job_status` before doing the merges

**Comment by denesb** [comment]
> v2:
> 
>     * move `check_jenkins_job_status` before doing the merges

Did you forget to force-push? I still see the check at the end.

**Comment by yaronkaikov** [comment]
> > v2:
> > ```
> > * move `check_jenkins_job_status` before doing the merges
> > ```
> 
> Did you forget to force-push? I still see the check at the end.

yes i did. :-) pushed now

**Review by denesb — APPROVED** [review]

**Comment by denesb** [comment]
Queued without CI, `scripts/pull_gh_pr.sh` is not covered by CI anyway.

**Comment by yaronkaikov** [comment]
> Queued without CI, `scripts/pull_github_pr.sh` is not covered by CI anyway.

I have aborted CI

File: scripts/pull_github_pr.sh
Comment (nyh) [reply scripts/pull_github_pr.sh]: I disagree with @denesb on this :-( I wish we discussed this.

As you can see We already have checks in the script - warning you if the tests didn't pass - and @yaronkaikov put his new check in exactly the same place. The **reason** why these checks are in that place (and last time, it was discussed) - is that maintainers like me sometimes want to try out a commit that they know is not ready. They want to pull it, but not push it. So the script does the pull, but then warns you not to push.

I guess that if you added an option ("--force") to do the operation anyway, then it's fine, but let's move all the tests to one place - it's makes little sense to test Jenkins in one place in a script and Github actions in a different place in the script.

Also, if you didn't do this (I didn't check), please make sure that the error message contains a mention of the option "--force" that you can use to ignore that check - when I need it, I won't remember how that option is called.

Comment (yaronkaikov) [reply scripts/pull_github_pr.sh]: @nyh I am about to send a follow-up PR with those changes (already talked about it with @denesb )

**Comment by xemul** [comment]
This broke the script to me @yaronkaikov 
```
$ sh -x ./scripts/pull_github_pr.sh 20675
+ set -e
+ gh_hosts=/home/xemul/.config/gh/hosts.yml
+ jenkins_url=https://jenkins.scylladb.com
+ [[ -z '' ]]
+ [[ -f /home/xemul/.config/gh/hosts.yml ]]
+ [[ 1 != 1 ]]
+ for required in jq curl
+ type jq
+ for required in jq curl
+ type curl
+ check_jenkins_job_status
+ set_jenkins_job
++ git rev-parse --abbrev-ref HEAD
+ branch=br-ernest-aws-errors
+ [[ br-ernest-aws-errors == \n\e\x\t ]]
+ [[ br-ernest-aws-errors == next-* ]]
+ echo 'Checking Jenkins job status for job: '
Checking Jenkins job status for job: 
+ lastCompletedBuild='https://jenkins.scylladb.com/job//lastCompletedBuild/api/json?tree=result'
++ curl -s --user : 'https://jenkins.scylladb.com/job//lastCompletedBuild/api/json?tree=result'
++ opts=()
++ local opts
++ [[ -n '' ]]
++ command curl -s --user : 'https://jenkins.scylladb.com/job//lastCompletedBuild/api/json?tree=result'
++ jq -r .result
jq: parse error: Invalid numeric literal at line 2, column 0
+ lastCompleted=
```

**Comment by yaronkaikov** [comment]
> merge script improvement, no need for backport

Sorry, you need to create a Jenkins token and set it as part of your env, i have sent a separate email about it. Also there is a follow-up PR https://github.com/scylladb/scylladb/pull/20793

**Comment by yaronkaikov** [comment]
v3:
- Script can run from any `next` branches
- Add an explanation about proper error messages when Jenkins credentials are missing
- Added option for `--force` when gating is unstable but we still want to merge

**Comment by yaronkaikov** [comment]
> v3:
> 
> * Script can run from any `next` branches
> * Add an explanation about proper error messages when Jenkins credentials are missing
> * Added option for `--force` when gating is unstable but we still want to merge

@nyh @denesb please review

File: scripts/pull_github_pr.sh
Comment (nyh) [inline scripts/pull_github_pr.sh:53]: This is not critical, but I want to point out that I use an old version of our pull_github_pr.sh not just in Scylla, but also in Seastar and even OSv (!) - it's a damn useful script (thanks @avikivity) to replace the nasty "merge" button on github. And this 100%-Scylla-specific code is breaking this usefulness. I would have split the script into two parts, but even if not, it would be nice to make *this* part more forgiving - i.e., if SCYLLA-VERSION-GEN you could just ignore all this Jenkins stuff, instead of failing in weird ways.

Oh, and instead of using cat grep **and** awk, you could do all of this with one command:
```
awk -F'=' '/^PRODUCT/ {print $2}' <SCYLLA-VERSION-GEN
```

File: scripts/pull_github_pr.sh
Comment (nyh) [inline scripts/pull_github_pr.sh:62]: Please consider using "case" instead of if - it's shorter, nicer, and the globbing is less suspicious:
```
case "$branch" in
next) jenkins_job=scylla-master/job/next;;
next-enterprise) jenkins_job=scylla-enterprise/job/next;;
next*) ...
   ;;;
esac
```

File: scripts/pull_github_pr.sh
Comment (nyh) [inline scripts/pull_github_pr.sh:78]: Please put the "*" inside the quotes, not outside. It looks like filename globbing this way (even if the bash [[ avoids the globbing, it still looks very weird).

**Review by nyh — APPROVED** [review]
Looks mostly good to me, I left some minor comments about improving the shell scripting style and avoiding wasteful or unnecessarily long constructs - but I think the functionality is correct now so I chose "approve".

File: scripts/pull_github_pr.sh
Comment (yaronkaikov) [reply scripts/pull_github_pr.sh:78]: Fixed

File: scripts/pull_github_pr.sh
Comment (yaronkaikov) [reply scripts/pull_github_pr.sh]: fixed

File: scripts/pull_github_pr.sh
Comment (yaronkaikov) [reply scripts/pull_github_pr.sh]: changed

**Comment by yaronkaikov** [comment]
> This broke the script to me @yaronkaikov
> 
> ```
> $ sh -x ./scripts/pull_github_pr.sh 20675
> + set -e
> + gh_hosts=/home/xemul/.config/gh/hosts.yml
> + jenkins_url=https://jenkins.scylladb.com
> + [[ -z '' ]]
> + [[ -f /home/xemul/.config/gh/hosts.yml ]]
> + [[ 1 != 1 ]]
> + for required in jq curl
> + type jq
> + for required in jq curl
> + type curl
> + check_jenkins_job_status
> + set_jenkins_job
> ++ git rev-parse --abbrev-ref HEAD
> + branch=br-ernest-aws-errors
> + [[ br-ernest-aws-errors == \n\e\x\t ]]
> + [[ br-ernest-aws-errors == next-* ]]
> + echo 'Checking Jenkins job status for job: '
> Checking Jenkins job status for job: 
> + lastCompletedBuild='https://jenkins.scylladb.com/job//lastCompletedBuild/api/json?tree=result'
> ++ curl -s --user : 'https://jenkins.scylladb.com/job//lastCompletedBuild/api/json?tree=result'
> ++ opts=()
> ++ local opts
> ++ [[ -n '' ]]
> ++ command curl -s --user : 'https://jenkins.scylladb.com/job//lastCompletedBuild/api/json?tree=result'
> ++ jq -r .result
> jq: parse error: Invalid numeric literal at line 2, column 0
> + lastCompleted=
> ```

@xemul Can you review the last changes?

**Review by nyh — APPROVED** [review]

**Comment by kbr-scylla** [comment]
Looks like there was a failure to test the script.... Again!
```
kbraun@fedora:~/dev/scylladb$ scripts/pull_github_pr.sh 20832
WARNING: https://jenkins.scylladb.com/job/scylla-master/job/next/lastCompletedBuild is not stable
exiting script since https://jenkins.scylladb.com/job/scylla-master/job/next/lastCompletedBuild is unstable, you can force it by running ./script/pull_github_pr.sh <pr num> --force
kbraun@fedora:~/dev/scylladb$ scripts/pull_github_pr.sh 20832 --force
jq: parse error: Invalid numeric literal at line 3, column 12
```

:/

**Comment by xemul** [comment]
Doesn't work again, @yaronkaikov 
```
$ ./scripts/pull_github_pr.sh 20730
jq: parse error: Invalid numeric literal at line 5, column 12
```

Detailed:

```
$ curl -s --user xemul:<token-here> 'https://jenkins.scylladb.com/job//lastCompletedBuild/api/json?tree=result' | jq -r .result
jq: parse error: Invalid numeric literal at line 5, column 12
```

```
$ curl -s --user xemul:<token-here> 'https://jenkins.scylladb.com/job//lastCompletedBuild/api/json?tree=result' 

    
    
  
  <!DOCTYPE html><html><head resURL="/static/c7380031" data-rooturl="" data-resurl="/static/c7380031" data-extensions-available="true" data-unit-test="false" data-imagesurl="/static/c7380031/images" data-crumb-header="Jenkins-Crumb" data-crumb-value="c5b17aff402067e94d8c597462c6839a3e3464f279f678c69990bbfaf172af2b">
    
    

    <title>Not Found [Jenkins]</title><link rel="stylesheet" href="/static/c7380031/jsbundles/styles.css" type="text/css"><link rel="stylesheet" href="/static/c7380031/css/responsive-grid.css" type="text/css"><link rel="icon" href="/static/c7380031/favicon.svg" type="image/svg+xml"><link sizes="any" rel="alternate icon" href="/static/c7380031/favicon.ico"><link sizes="180x180" rel="apple-touch-icon" href="/static/c7380031/apple-touch-icon.png"><link color="#191717" rel="mask-icon" href="/static/c7380031/mask-icon.svg"><meta name="theme-color" content="#ffffff"><script src="/static/c7380031/scripts/behavior.js" type="text/javascript"></script><script src='/adjuncts/c7380031/org/kohsuke/stapler/bind.js' type='text/javascript'></script><script src="/static/c7380031/scripts/yui/yahoo/yahoo-min.js"></script><script src="/static/c7380031/scripts/yui/dom/dom-min.js"></script><script src="/static/c7380031/scripts/yui/event/event-min.js"></script><script src="/static/c7380031/scripts/yui/animation/animation-min.js"></script><script src="/static/c7380031/scripts/yui/dragdrop/dragdrop-min.js"></script><script src="/static/c7380031/scripts/yui/container/container-min.js"></script><script src="/static/c7380031/scripts/yui/connection/connection-min.js"></script><script src="/static/c7380031/scripts/yui/datasource/datasource-min.js"></script><script src="/static/c7380031/scripts/yui/autocomplete/autocomplete-min.js"></script><script src="/static/c7380031/scripts/yui/menu/menu-min.js"></script><script src="/static/c7380031/scripts/yui/element/element-min.js"></script><script src="/static/c7380031/scripts/yui/button/button-min.js"></script><script src="/static/c7380031/scripts/yui/storage/storage-min.js"></script><script src="/static/c7380031/scripts/hudson-behavior.js" type="text/javascript"></script><script src="/static/c7380031/scripts/sortable.js" type="text/javascript"></script><link rel="stylesheet" href="/static/c7380031/scripts/yui/container/assets/container.css" type="text/css"><link rel="stylesheet" href="/static/c7380031/scripts/yui/container/assets/skins/sam/container.css" type="text/css"><link rel="stylesheet" href="/static/c7380031/scripts/yui/menu/assets/skins/sam/menu.css" type="text/css"><link rel="search" href="/opensearch.xml" type="application/opensearchdescription+xml" title="Jenkins"><meta name="ROBOTS" content="INDEX,NOFOLLOW"><meta name="viewport" content="width=device-width, initial-scale=1"><script src="/adjuncts/c7380031/org/kohsuke/stapler/jquery/jquery.full.js" type="text/javascript"></script><script>var Q=jQuery.noConflict()</script><link rel="stylesheet" href="/plugin/jquery-ui/css/jquery-ui-1.8.9.custom.css" type="text/css"><script src="/plugin/jquery-ui/js/jquery-ui-1.8.9.custom.min.js"></script><script src='/adjuncts/c7380031/org/jenkinsci/plugins/scriptsecurity/scripts/ScriptApproval/FormValidationPageDecorator/validate.js' type='text/javascript'></script><script src="/static/c7380031/jsbundles/vendors.js" type="text/javascript"></script><script src="/static/c7380031/jsbundles/sortable-drag-drop.js" type="text/javascript"></script><script defer="true" src="/static/c7380031/jsbundles/app.js" type="text/javascript"></script></head><body data-model-type="hudson.model.Hudson" id="jenkins" class="yui-skin-sam one-column jenkins-2.461" data-version="2.461"><a href="#skip2content" class="jenkins-skip-link">Skip to content</a><header id="page-header" class="page-header"><div class="page-header__brand"><div class="logo"><a id="jenkins-home-link" href="/"><img src="/static/c7380031/images/svgs/logo.svg" alt="[Jenkins]" id="jenkins-head-icon"><img src="/static/c7380031/images/title.svg" alt="Jenkins" width="139" id="jenkins-name-icon" height="34"></a></div><a href="/" class="page-header__brand-link"><img src="/static/c7380031/images/svgs/logo.svg" alt="[Jenkins]" class="page-header__brand-image"><span class="page-header__brand-name">Jenkins</span></a></div><div class="searchbox hidden-xs"><form role="search" method="get" name="search" action="/search/" style="position:relative;" class="no-json"><div id="search-box-sizer"></div><div id="searchform"><input role="searchbox" name="q" placeholder="Search" id="search-box" class="main-search__input"><span class="main-search__icon-leading"><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" viewBox="0 0 512 512"><title/><path d="M221.09 64a157.09 157.09 0 10157.09 157.09A157.1 157.1 0 00221.09 64z" fill="none" stroke="currentColor" stroke-miterlimit="10" stroke-width="32"/><path d="M338.29 338.29L448 448" fill="none" stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="32"/></svg></span><a href="https://www.jenkins.io/redirect/search-box" class="main-search__icon-trailing"><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" viewBox="0 0 512 512"><path d="M256 40a216 216 0 10216 216A216 216 0 00256 40z" fill="none" stroke="currentColor" stroke-miterlimit="10" stroke-width="38"/><path d="M200 202.29s.84-17.5 19.57-32.57C230.68 160.77 244 158.18 256 158c10.93-.14 20.69 1.67 26.53 4.45 10 4.76 29.47 16.38 29.47 41.09 0 26-17 37.81-36.37 50.8S251 281.43 251 296" fill="none" stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="38"/><circle cx="250" cy="360" fill="currentColor" r="25"/></svg></a><div id="search-box-completion" data-search-url="/search/"></div><script src='/adjuncts/c7380031/jenkins/views/JenkinsHeader/search-box.js' type='text/javascript'></script></div></form></div><script src="/static/c7380031/jsbundles/keyboard-shortcuts.js" type="text/javascript"></script><div class="login page-header__hyperlinks"><div id="visible-am-insertion" class="page-header__am-wrapper"></div><div id="visible-sec-am-insertion" class="page-header__am-wrapper"></div><a href="/user/xemul" class="model-link"><svg class="icon-md" xmlns="http://www.w3.org/2000/svg" aria-hidden="true" viewBox="0 0 512 512"><title/><path d="M258.9 48C141.92 46.42 46.42 141.92 48 258.9c1.56 112.19 92.91 203.54 205.1 205.1 117 1.6 212.48-93.9 210.88-210.88C462.44 140.91 371.09 49.56 258.9 48zm126.42 327.25a4 4 0 01-6.14-.32 124.27 124.27 0 00-32.35-29.59C321.37 329 289.11 320 256 320s-65.37 9-90.83 25.34a124.24 124.24 0 00-32.35 29.58 4 4 0 01-6.14.32A175.32 175.32 0 0180 259c-1.63-97.31 78.22-178.76 175.57-179S432 158.81 432 256a175.32 175.32 0 01-46.68 119.25z" fill="currentColor"/><path d="M256 144c-19.72 0-37.55 7.39-50.22 20.82s-19 32-17.57 51.93C191.11 256 221.52 288 256 288s64.83-32 67.79-71.24c1.48-19.74-4.8-38.14-17.68-51.82C293.39 151.44 275.59 144 256 144z" fill="currentColor"/></svg><span class="hidden-xs hidden-sm">Pavel Emelyanov</span></a><a href="/logout"><svg class="icon-md" xmlns="http://www.w3.org/2000/svg" aria-hidden="true" viewBox="0 0 512 512"><title/><path d="M304 336v40a40 40 0 01-40 40H104a40 40 0 01-40-40V136a40 40 0 0140-40h152c22.09 0 48 17.91 48 40v40M368 336l80-80-80-80M176 256h256" fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="32"/></svg><span class="hidden-xs hidden-sm">log out</span></a></div></header><div id="breadcrumbBar" class="jenkins-breadcrumbs" aria-label="breadcrumb"><ol class="jenkins-breadcrumbs__list" id="breadcrumbs"><li class="jenkins-breadcrumbs__list-item"><a href="/" class="model-link">Dashboard</a></li><li class="children" data-href="/"></li></ol></div><div id="page-body" class="app-page-body app-page-body--one-column clear"><div id="main-panel"><a id="skip2content"></a><h1 style="text-align: center"><img src="/static/c7380031/images/rage.svg" width="154" height="179"><span style="font-size:50px"> Oops!</span></h1><div id="error-description" style="text-align: center"><h2>Not Found</h2><p>This page may not exist, or you may not have permission to see it.</div></div></div><footer class="page-footer jenkins-mobile-hide"><div class="page-footer__flex-row"><div class="page-footer__footer-id-placeholder" id="footer"></div><div class="page-footer__links"><button type="button" class="jenkins-button jenkins-button--tertiary jenkins_ver" data-dropdown="true">

    Jenkins 2.461
  </button><template><div class="jenkins-dropdown"><template data-dropdown-icon="&lt;svg xmlns=&quot;http://www.w3.org/2000/svg&quot; aria-hidden=&quot;true&quot; viewBox=&quot;0 0 512 512&quot;&gt;&lt;path d=&quot;M352.92 80C288 80 256 144 256 144s-32-64-96.92-64c-52.76 0-94.54 44.14-95.08 96.81-1.1 109.33 86.73 187.08 183 252.42a16 16 0 0018 0c96.26-65.34 184.09-143.09 183-252.42-.54-52.67-42.32-96.81-95.08-96.81z&quot; fill=&quot;none&quot; stroke=&quot;currentColor&quot; stroke-linecap=&quot;round&quot; stroke-linejoin=&quot;round&quot; stroke-width=&quot;32&quot;/&gt;&lt;/svg&gt;" data-dropdown-text="Get involved" data-dropdown-type="ITEM" data-dropdown-href="https://www.jenkins.io/participate/"></template><template data-dropdown-icon="&lt;svg xmlns=&quot;http://www.w3.org/2000/svg&quot; aria-hidden=&quot;true&quot; viewBox=&quot;0 0 512 512&quot;&gt;&lt;path d=&quot;M384 224v184a40 40 0 01-40 40H104a40 40 0 01-40-40V168a40 40 0 0140-40h167.48M336 64h112v112M224 288L440 72&quot; fill=&quot;none&quot; stroke=&quot;currentColor&quot; stroke-linecap=&quot;round&quot; stroke-linejoin=&quot;round&quot; stroke-width=&quot;32&quot;/&gt;&lt;/svg&gt;" data-dropdown-text="Website" data-dropdown-type="ITEM" data-dropdown-href="https://www.jenkins.io/"></template></div></template></div></div></footer></body></html>
```

**Comment by xemul** [comment]
Reverted again (c17d353718464bae230331d7e77d4901281fc017)

**Comment by yaronkaikov** [comment]
> 20832

@xemul could it be that you didn't run the script from the `next` branch? your command is missing the job name `curl -s --user xemul:<token-here> 'https://jenkins.scylladb.com/job//lastCompletedBuild/api/json?tree=result' ` it should be `curl -s --user xemul:<token-here> 'https://jenkins.scylladb.com/job/next/lastCompletedBuild/api/json?tree=result' 

I have just checked it again and it's working as expected

**Comment by xemul** [comment]
> @xemul could it be that you didn't run the script from the next branch? 

I always run it on the next branch, this script is pointless for any other branch (except for next-$version, but that's also next)

**Comment by yaronkaikov** [comment]
v4:

- Notify the user if trying to run the script from branches other than `next`
- Reduce Jenkins API calls to single call

**Comment by yaronkaikov** [comment]
> v4:
> 
> * Notify the user if trying to run the script from branches other than `next`
> * Reduce Jenkins API calls to a single call

@kbr-scylla @xemul Could you please verify this change locally? I did verification locally, but let's make sure :-)

**Comment by nyh** [comment]
> v4:
> 
>     * Notify the user if trying to run the script from branches other than `next`

Please note that I always run this script not just from next, but also "next-*", "next-enterprise-*", etc.

By the way, as I noted in the past, I use an **old** version of this script also in the Seastar project, and various other Scylla projects (even OSv). The new versions added a lot of super-CI-specific stuff that no longer works, but the original script was a gem, and I use in various places.

**Comment by denesb** [comment]
@yaronkaikov I think you need to make this new code more resilient against random Jenkins flukes and failures.
At the very least `--force` should allow a maintainer to merge a PR, regardless of in what way the Jenkins request failed. Currently, if Jenkins replies with some HTML response, the script will fail and `--force` will not help.

**Comment by denesb** [comment]
This might require rewriting this script in python first... bash has horrible error handling.

**Comment by yaronkaikov** [comment]
> > v4:
> > ```
> > * Notify the user if trying to run the script from branches other than `next`
> > ```
> 
> Please note that I always run this script not just from next, but also "next-_", "next-enterprise-_", etc.
> 
> By the way, as I noted in the past, I use an **old** version of this script also in the Seastar project, and various other Scylla projects (even OSv). The new versions added a lot of super-CI-specific stuff that no longer works, but the original script was a gem, and I use in various places.

yes it allow to run from `next` , `next-enterprise` and `next-*` , i wasn't clear enough

**Comment by yaronkaikov** [comment]
> @yaronkaikov I think you need to make this new code more resilient against random Jenkins flukes and failures. At the very least `--force` should allow a maintainer to merge a PR, regardless of in what way the Jenkins request failed. Currently, if Jenkins replies with some HTML response, the script will fail and `--force` will not help.

v5:
- use `--force` to skip the Jenkin check completely

**Comment by nyh** [comment]
> This might require rewriting this script in python first... bash has horrible error handling.

As they say, Bash is user-friendly - it's just picky about its friends :-)

**Review by denesb — APPROVED** [review]

**Comment by denesb** [comment]
> [test_tablet_drain_failure_during_decommission.dev.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12030/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission_dev_4) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission.dev.4)

https://github.com/scylladb/scylladb/issues/20060

**Pull request closed without merging**
